mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py

```diff
@@ -0,0 +1,397 @@
+        """
+        assert metric == 'mAP'
+
+        res_file = os.path.join(res_folder, 'result_keypoints.json')
+
+        kpts = defaultdict(list)
+        for preds, boxes, image_path in outputs:
+            str_image_path = ''.join(image_path)
+            image_id = int(str_image_path[-10:-4])
+
+            kpts[image_id].append({
+                'keypoints': preds[0],
+                'center': boxes[0][0:2],
+                'scale': boxes[0][2:4],
+                'area': boxes[0][4],
+                'score': boxes[0][5],
+                'image_id': image_id,
+            })
+
+        # rescoring and oks nms
+        num_joints = self.ann_info['num_joints']
+        vis_thr = self.vis_thr
+
+        final_result = []
+        for img in kpts.keys():
+            img_kpts = kpts[img]
+            boxes, preds, box_scores = convert_crowd(img_kpts)
+            result = candidate_reselect(boxes, preds, num_joints, img,
+                                        box_scores, vis_thr)
+            final_result.append(result)
+
+        self._write_coco_keypoint_results(final_result, res_file)
+
+        info_str = self._do_python_keypoint_eval(res_file)
+        name_value = OrderedDict(info_str)
+
+        return name_value
+
+    def _write_coco_keypoint_results(self, keypoints, res_file):
+        """Write results into a json file."""
+        data_pack = [{
+            'cat_id': self._class_to_coco_ind[cls],
+            'cls_ind': cls_ind,
+            'cls': cls,
+            'ann_type': 'keypoints',
+            'keypoints': keypoints
+        } for cls_ind, cls in enumerate(self.classes)
+                     if not cls == '__background__']
+
+        results = self._coco_keypoint_results_one_category_kernel(data_pack[0])
+
+        with open(res_file, 'w') as f:
+            json.dump(results, f, sort_keys=True, indent=4)
+
+    def _coco_keypoint_results_one_category_kernel(self, data_pack):
+        """Get crowdpose keypoint results."""
+        cat_id = data_pack['cat_id']
+        keypoints = data_pack['keypoints']
+        cat_results = []
+
+        for img_kpts in keypoints:
+            if len(img_kpts) == 0:
+                continue
+
+            _key_points = np.array(
+                [img_kpt['keypoints'] for img_kpt in img_kpts])
+            key_points = _key_points.reshape(-1,
+                                             self.ann_info['num_joints'] * 3)
+
+            result = [{
+                'image_id': img_kpt['image_id'],
+                'category_id': cat_id,
+                'keypoints': list(key_point),
+                'score': img_kpt['score'],
+                'center': list(img_kpt['center']),
+                'scale': list(img_kpt['scale'])
+            } for img_kpt, key_point in zip(img_kpts, key_points)]
+
+            cat_results.extend(result)
+
+        return cat_results
+
+    def _do_python_keypoint_eval(self, res_file):
```

Comment:
does "python" have special meaning?

Reply:
no. just named the same as  coco dataset's keypoint eval

Reply:
okay we can give it a better name